homeassistant/components/mqtt/humidifier.py

```diff
@@ -0,0 +1,488 @@
+            vol.Optional(
+                CONF_TARGET_HUMIDITY_MIN, default=DEFAULT_MIN_HUMIDITY
+            ): cv.positive_int,
+            vol.Optional(
+                CONF_TARGET_HUMIDITY_MAX, default=DEFAULT_MAX_HUMIDITY
+            ): cv.positive_int,
+            vol.Optional(CONF_TARGET_HUMIDITY_RANGE_MIN): cv.positive_int,
+            vol.Optional(CONF_TARGET_HUMIDITY_RANGE_MAX): cv.positive_int,
+            vol.Optional(
+                CONF_PAYLOAD_RESET_HUMIDITY, default=DEFAULT_PAYLOAD_RESET
+            ): cv.string,
+            vol.Optional(
+                CONF_PAYLOAD_RESET_MODE, default=DEFAULT_PAYLOAD_RESET
+            ): cv.string,
+            vol.Optional(CONF_PAYLOAD_OFF, default=DEFAULT_PAYLOAD_OFF): cv.string,
+            vol.Optional(CONF_PAYLOAD_ON, default=DEFAULT_PAYLOAD_ON): cv.string,
+            vol.Optional(CONF_STATE_VALUE_TEMPLATE): cv.template,
+        }
+    ).extend(MQTT_ENTITY_COMMON_SCHEMA.schema),
+    valid_humidity_range_configuration,
+    valid_mode_configuration,
+)
+
+
+async def async_setup_platform(
+    hass: HomeAssistant, config: ConfigType, async_add_entities, discovery_info=None
+):
+    """Set up MQTT humidifier through configuration.yaml."""
+    await async_setup_reload_service(hass, DOMAIN, PLATFORMS)
+    await _async_setup_entity(hass, async_add_entities, config)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up MQTT humidifier dynamically through MQTT discovery."""
+    setup = functools.partial(
+        _async_setup_entity, hass, async_add_entities, config_entry=config_entry
+    )
+    await async_setup_entry_helper(hass, humidifier.DOMAIN, setup, PLATFORM_SCHEMA)
+
+
+async def _async_setup_entity(
+    hass, async_add_entities, config, config_entry=None, discovery_data=None
+):
+    """Set up the MQTT humidifier."""
+    async_add_entities([MqttHumidifier(hass, config, config_entry, discovery_data)])
+
+
+class MqttHumidifier(MqttEntity, HumidifierEntity):
+    """A MQTT humidifier component."""
+
+    _attributes_extra_blocked = MQTT_HUMIDIFIER_ATTRIBUTES_BLOCKED
+
+    def __init__(self, hass, config, config_entry, discovery_data):
+        """Initialize the MQTT humidifier."""
+        self._state = False
+        self._target_humidity = None
+        self._mode = None
+        self._supported_features = 0
+
+        self._topic = None
+        self._payload = None
+        self._value_templates = None
+        self._command_templates = None
+        self._optimistic = None
+        self._optimistic_target_humidity = None
+        self._optimistic_mode = None
+
+        MqttEntity.__init__(self, hass, config, config_entry, discovery_data)
+
+    @staticmethod
+    def config_schema():
+        """Return the config schema."""
+        return PLATFORM_SCHEMA
+
+    def _setup_from_config(self, config):
+        """(Re)Setup the entity."""
+        self._attr_device_class = DEVICE_CLASS_HUMIDIFIER
```

Comment:
Can we add this to the configuration schema, we should allow either `humidifier` or `dehumidifier`

Reply:
Added this to the config schema. Have a look at the validation method. I used a regular expression.